Geometry_Engine/Query/GeometryHash.cs

```diff
@@ -0,0 +1,427 @@
+        }
+
+        /***************************************************/
+
+        [Description("The GeometryHash for an Circle is calculated as the GeometryHash of the start, 1/3rd and 2/3rd points of the Circle.")]
+        private static double[] GeometryHash(this Circle curve, double translationFactor, bool skipLastPoint = false)
+        {
+            translationFactor += (int)TypeTranslationFactor.Circle;
+
+            IEnumerable<double> hash = curve.StartPoint().ToDoubleArray(translationFactor)
+                   .Concat(curve.PointAtParameter(0.33).ToDoubleArray(translationFactor));
+
+            if (!skipLastPoint)
+                hash = hash.Concat(curve.PointAtParameter(0.66).ToDoubleArray(translationFactor));
+
+            return hash.ToArray();
+        }
+
+        /***************************************************/
+
+        [Description("The GeometryHash for an Ellipse is calculated as the GeometryHash of the start, 1/3rd and 2/3rd points of the Ellipse.")]
+        private static double[] GeometryHash(this Ellipse curve, double translationFactor, bool skipLastPoint = false)
+        {
+            translationFactor += (int)TypeTranslationFactor.Ellipse;
+
+            IEnumerable<double> hash = curve.StartPoint().ToDoubleArray(translationFactor)
+               .Concat(curve.PointAtParameter(0.33).ToDoubleArray(translationFactor));
+
+            if (!skipLastPoint)
+                hash = hash.Concat(curve.PointAtParameter(0.66).ToDoubleArray(translationFactor));
+
+            return hash.ToArray();
+        }
+
+        /***************************************************/
+
+        [Description("The GeometryHash for a Line is calculated as the GeometryHash of the start and end point of the Line.")]
+        private static double[] GeometryHash(this Line curve, double translationFactor, bool skipLastPoint = false)
+        {
+            translationFactor += (int)TypeTranslationFactor.Line;
+
+            if (skipLastPoint)
+                return curve.StartPoint().ToDoubleArray(translationFactor);
+
+            return curve.StartPoint().ToDoubleArray(translationFactor)
+               .Concat(curve.EndPoint().ToDoubleArray(translationFactor))
+               .ToArray();
+        }
+
+        /***************************************************/
+
+        [Description("The GeometryHash for a NurbsCurve is obtained by getting moving the control points " +
+            "by a translation factor composed by the weights and a subarray of the knot vector. " +
+            "The subarray is made by picking as many elements from the knot vector as the curve degree value.")]
+        private static double[] GeometryHash(this NurbsCurve curve, double translationFactor, bool skipLastPoint = false)
+        {
+            int curveDegree = curve.Degree();
+
+            if (curveDegree == 1)
+                return BH.Engine.Geometry.Create.Polyline(curve.ControlPoints).GeometryHash(translationFactor);
```

Comment:
This there could be an edge case of an edgecase that could happen here when a nurbscurve of degree 1 is part of a polycurve, and then if that nurbscurve is changed to a polyline, then the subparts might start to act slightly differently with regards to the skip last point.

This can potentially be fixed by adding a SubParts method for the NurbsCurve, and return a list of Lines corresponding to the Polyline if the curve is Degree 1.

Do not think this is anything that needs to be handled in this PR, again mostly noting in case I read this again.